engine/src/main/java/com/google/android/fhir/sync/Sync.kt

```diff
@@ -147,14 +248,95 @@ object Sync {
         it.backoffCriteria.timeUnit,
       )
       periodicWorkRequestBuilder.setInputData(
-        Data.Builder().putInt(MAX_RETRIES_ALLOWED, it.maxRetries).build(),
+        Data.Builder()
+          .putInt(MAX_RETRIES_ALLOWED, it.maxRetries)
+          .putString(SYNC_STATUS_PREFERENCES_DATASTORE_KEY, uniqueWorkName)
+          .build(),
       )
     }
     return periodicWorkRequestBuilder.build()
   }
 
   /** Gets the timestamp of the last sync job. */
   fun getLastSyncTimestamp(context: Context): OffsetDateTime? {
-    return DatastoreUtil(context).readLastSyncTimestamp()
+    return FhirEngineProvider.getFhirDataStore(context).readLastSyncTimestamp()
   }
+
+  private fun createSyncState(
+    workInfoState: WorkInfo.State,
+    currentSyncJobStatus: SyncJobStatus,
+    terminalSyncJobStatus: SyncJobStatus,
+  ): SyncState {
+    return when (workInfoState) {
+      ENQUEUED -> {
+        Enqueued
+      }
+      RUNNING -> {
+        Running(currentSyncJobStatus)
+      }
+      SUCCEEDED -> {
+        Succeeded(terminalSyncJobStatus)
+      }
+      FAILED -> {
+        Failed(terminalSyncJobStatus)
+      }
+      CANCELLED -> {
+        Cancelled
+      }
+      else -> {
+        Unknown
+      }
+    }
+  }
```

Comment:
I think how i've written it simplifies the code and is based on the table that you had drawn in your design doc
```suggestion
  private fun createSyncState(
    workInfoState: WorkInfo.State,
    syncJobStatusFromWorkManager: SyncJobStatus?,
    syncJobStatusFromDataStore: SyncJobStatus?,
  ): SyncState {
    return when (syncJobStatusFromWorkManager) {
      is SyncJobStatus.Started,
      is SyncJobStatus.InProgress, -> Running(syncJobStatusFromWorkManager)
      is SyncJobStatus.Finished -> Succeeded(syncJobStatusFromDataStore!!)
      is SyncJobStatus.Failed -> Failed(syncJobStatusFromDataStore!!)
      null -> {
        when (workInfoState) {
          RUNNING -> Running(SyncJobStatus.Started())
          ENQUEUED -> Enqueued
          CANCELLED -> Cancelled
          SUCCEEDED -> Succeeded(SyncJobStatus.Finished())
          else -> error("Error message here")
        }
      }
    }
  }
```

Reply:
"Yes, it simplifies the code. Thanks for that!"
These are my observations:

1. Below cases will never be executed, as the terminal state involves only writing and reading from the dataStore. However, I understand that to avoid the 'else' case, we have explicitly mentioned those states here, correct?
```
is SyncJobStatus.Finished -> Succeeded(syncJobStatusFromDataStore!!)
is SyncJobStatus.Failed -> Failed(syncJobStatusFromDataStore!!)
```

2. Is the terminal state of syncJobStatus in periodic sync only updated through `lastJobState`? So, to determine whether the current job is finished or failed, it is required to read `PeriodicSyncState (lastJobState = Failed/Succeeded, currentJobState = Enqueued)`. One property is not enough, correct?

3. In OneTimeSync, if terminal state is failed, then it does not throw exceptions from SyncJobStatus.Failed

Reply:
> "Yes, it simplifies the code. Thanks for that!" These are my observations:
> 
> 1. Below cases will never be executed, as the terminal state involves only writing and reading from the dataStore. However, I understand that to avoid the 'else' case, we have explicitly mentioned those states here, correct?
> 
> ```
> is SyncJobStatus.Finished -> Succeeded(syncJobStatusFromDataStore!!)
> is SyncJobStatus.Failed -> Failed(syncJobStatusFromDataStore!!)
> ```

This is an excellent observation that I didn't realise. I tested for a successful sync and you're right, `SyncJobStatus.Finished` is never executed ! We probably need to handle the case where `syncJobStatusInDataStoreFlow` state is Failed like this though:

```
  private fun createSyncState(
    workInfoState: WorkInfo.State,
    syncJobStatusFromWorkManager: SyncJobStatus?,
    syncJobStatusFromDataStore: SyncJobStatus?,
  ): SyncState =
     when (syncJobStatusFromWorkManager) {
      is SyncJobStatus.Started,
      is SyncJobStatus.InProgress -> Running(syncJobStatusFromWorkManager)
      null -> {
        syncJobStatusFromDataStore?.let {
          return when (it) {
            is SyncJobStatus.Failed -> SyncState.Failed(it)
            is SyncJobStatus.Finished -> Succeeded(it)
            else -> error("Error message here")
          }
        }
        when (workInfoState) {
          RUNNING -> Running(SyncJobStatus.Started())
          ENQUEUED -> Enqueued
          CANCELLED -> Cancelled
          SUCCEEDED -> Succeeded(SyncJobStatus.Finished())
          else -> error("Error message here")
        }
      }
      else -> error("Error message here")
    }
```

WDYT?

> 
> 2. Is the terminal state of syncJobStatus in periodic sync only updated through `lastJobState`? So, to determine whether the current job is finished or failed, it is required to read `PeriodicSyncState (lastJobState = Failed/Succeeded, currentJobState = Enqueued)`. One property is not enough, correct?
I think so yes

> 3. In OneTimeSync, if terminal state is failed, then it does not throw exceptions from SyncJobStatus.Failed

We pass the exceptions to the user

Reply:
yes, it looks ok to me, thanks! 
I just added small change to accommodate  ```(workInfoState= ENQUEUED, syncJobStatusFromWorkManager=null, syncJobStatusFromDataStore=Finished/Failed) ```


```
private fun createSyncState(
    workInfoState: WorkInfo.State,
    syncJobStatusFromWorkManager: SyncJobStatus?,
    syncJobStatusFromDataStore: SyncJobStatus?,
  ): SyncState =
    when (syncJobStatusFromWorkManager) {
      is SyncJobStatus.Started,
      is SyncJobStatus.InProgress, -> Running(syncJobStatusFromWorkManager)
      null -> {
        when (workInfoState) {
          RUNNING -> Running(SyncJobStatus.Started())
          ENQUEUED -> Enqueued
          CANCELLED -> Cancelled
          SUCCEEDED -> {
            syncJobStatusFromDataStore?.let {
              return when (it) {
                is SyncJobStatus.Finished -> Succeeded(it)
                else -> error("Error message here")
              }
            }
            Succeeded(SyncJobStatus.Finished())
          }
          FAILED -> {
            syncJobStatusFromDataStore?.let {
              return when (it) {
                is SyncJobStatus.Failed -> Failed(it)
                else -> error("Error message here")
              }
            }
            error("Error message here")
          }
          else -> error("Error message here")
        }
      }
      else -> error("Error message here")
    }
```

Reply:
nice! i refactored it a bit to to reduce the nested blocks and make it a bit easier to read:
 
```
private fun createSyncState(
    workInfoState: WorkInfo.State,
    syncJobStatusFromWorkManager: SyncJobStatus?,
    syncJobStatusFromDataStore: SyncJobStatus?,
): SyncState =
    when (syncJobStatusFromWorkManager) {
        is SyncJobStatus.Started,
        is SyncJobStatus.InProgress -> Running(syncJobStatusFromWorkManager)
        null -> handleNullWorkManagerStatus(workInfoState, syncJobStatusFromDataStore)
        else -> error("Error message here")
    }

private fun handleNullWorkManagerStatus(
    workInfoState: WorkInfo.State,
    syncJobStatusFromDataStore: SyncJobStatus?
): SyncState = when (workInfoState) {
    RUNNING -> Running(SyncJobStatus.Started())
    ENQUEUED -> Enqueued
    CANCELLED -> Cancelled
    SUCCEEDED, FAILED -> handleFinishedOrFailedState(workInfoState, syncJobStatusFromDataStore)
    else -> error("Error message here")
}

private fun handleFinishedOrFailedState(
    workInfoState: WorkInfo.State,
    syncJobStatusFromDataStore: SyncJobStatus?
): SyncState = syncJobStatusFromDataStore?.let {
    when (it) {
        is SyncJobStatus.Finished ->  Succeeded(it)
        is SyncJobStatus.Failed -> Failed(it)
        else -> error("Error message here")
    }
} ?: when (workInfoState) {
    SUCCEEDED -> Succeeded(SyncJobStatus.Finished())
    FAILED -> error("Error message here")
    else -> error("Error message here")
}
```

Reply:
done.